rtf_vis_tool/src/Components/frustum_classes/view_frustum.js

```diff
@@ -0,0 +1,168 @@
+        */
+
+        // Assume principal point is at center of images.
+        var px = img_w / 2;
+        var py = img_h / 2;
+
+        //uv - [px,py] gives each vertex's pixel offset from the center of image plane.
+        var center_offsets = nj.zeros([5,2]);
+        for (var row = 0; row < 5; row++) {
+            center_offsets.set(row, 0, uv.get(row, 0) - px);
+            center_offsets.set(row, 1, uv.get(row, 1) - py);
+        }
+
+        var fx_broadcasted = nj.array([
+            [this.fx_],
+            [this.fx_],
+            [this.fx_],
+            [this.fx_],
+            [this.fx_]
+        ]);
+        var ray_dirs = nj.concatenate(center_offsets, fx_broadcasted);
+        ray_dirs = this.normalize_ray_dirs(ray_dirs);
+        return ray_dirs
+    }
+
+    get_frustum_vertices_camfr() {
+        /*Obtain 3d positions of all 5 frustum vertices in the camera frame
+
+          (x,y,z)               (x,y,z)                (x,y,z)              (x,y,z)
+              \\=================//                      \\                   //
+               \\               //                        \\ 1-------------2 //
+        (-w/2,-h/2,fx)       (w/2,-h/2,fx)                 \\| IMAGE PLANE |//
+                 1-------------2                             |             |/
+                 |\\         //| IMAGE PLANE  (-w/2, h/2,fx) 4-------------3 (w/2, h/2,fx)
+                 | \\       // | IMAGE PLANE                  \\         //
+                 4--\\-----//--3                               \\       //
+                     \\   //                                    \\     //
+                      \\ //                                      \\   //
+                        O PINHOLE                                 \\ //
+                                                                    O PINHOLE
+            
+            Returns:
+                Array, size 5, of frustum vertex coordinates in the camera coordinate system.                                                                
+        */
+
+        var uv = nj.array([
+            [Math.floor(this.img_w_ / 2), Math.floor(this.img_h_ / 2)],  //v0 = optical center
+            [0, 0],                              //v1 = top-left
+            [this.img_w_ - 1, 0],                //v2 = top-right
+            [this.img_w_ - 1, this.img_h_ - 1],  //v3 = bottom-right
+            [0, this.img_h_ - 1],                //v4 = bottom-left
+        ]);
+
+        var ray_dirs = this.compute_pixel_ray_directions_vectorized(uv, this.img_w_, this.img_h_);
+        var scaled_ray_dirs = this.calculate_scaled_ray_dirs(ray_dirs, this.frustum_ray_len_);
+        return scaled_ray_dirs
+    }
+
+    get_mesh_vertices_worldframe(wTc) {
+        /*Obtains 5 vertices defining the frustum mesh, in the world/global frame.
+
+            Args:
+                wTc: camera pose in world frame
+            Returns:
+                An array, length 5, of the frustum vertices.
```

Comment:
length 5 or 5x3?

Reply:
or i guess "An array, length 5, where each entry represents the 3d coordinate of a frustum vertex"